### ISSUE TITLE
X509_dup() does not copy libctx (and likely neither any other non-ASN.1 fields)

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#12680]
One would expect that `X509_dup()` copies all fields, but at least for `libctx` this is not the case,
likely because it is implemented solely by a generic ASN.1 structure copying function 
and this does not copy other other fields of `x509_st`.

See also https://github.com/openssl/openssl/pull/11808#discussion_r471092440 and https://github.com/openssl/openssl/pull/11808#discussion_r472908941.

**Comment by slontis** [comment]
That is not good.. I assume this is not used very often then.. libctx needs to be copied.

**Comment by levitte** [comment]
I believe the callback should take care of that. There should be a callback operator for duplication...

**Comment by DDvO** [comment]
I cannot find any callback, at least not in `crypto/asn1/a_dup.c`.

**Comment by levitte** [comment]
Yup, I found that too...  so I'd suggest adding a duplication callback operation...  ASN1_OP_DUP_PRE and ASN1_OP_DUP_POST, say?

_PRE would be called on the source object, and _POST would be called on the destination object, *with the source object as last argument*.  That way, the callback can be used to prepare the source for duplication (if ever needed), and to copy extra stuff from the source to the destination.

Does that sound workable?  Are you willing to do it, @DDvO?  I've got my hands full already, so would rather not.
